### PR TITLE
Making a flag option in the data generator for generating a single target in an open receptacle

### PR DESCRIPTION
**Pull request by vincentpierre**
## Types of changes

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: habitat/datasets/rearrange/rearrange_generator.py
Comment (aclegg3) [inline habitat/datasets/rearrange/rearrange_generator.py:886]: Thanks for adding the comments. 
Seems like the edge case, modified behavior should not be default.
```suggestion
    #TODO: consider refactoring the generator to support this option more generically.
    _C.unique_target = False
```

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:346]:
```suggestion
            # To ensure at least one object will be sampled in the target_receptacle,
```

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:577]: I understand why this hardcoded override is necessary with current metadata, but I'm hesitant to merge it. I'd be more comfortable with a metadata solution which could be applied generically for other cases like this. Example: a receptacle inside the ReplicaCAD cabinet.

One idea, we could annotate receptacles with an additional optional field for link state requirements to access the space. In this paradigm, the fridge receptacles could be annotated with an acceptable joint range for the doors which allow access to them. Also note that this would be a very easy change since AO Receptacle metadata are defined with the URDF rather than distributed across the scene and stage configs.

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:586]: This assumes 0 joint state is default. This isn't necessarily the case for generic URDF. Since we allow state overrides in the scene config, it would be best to use the current state of joints upon initialization.

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:699]: same comment about defaults

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:690]: same comment about metadata

**Review by aclegg3 — COMMENTED** [review]
Thanks for iterating on this change and breaking it out into a separate PR. This is better than before, and I'd be ok with merging to unblock in the short term if necessary with some modifications. 

I still believe in refactoring to support generic, conditional AO sampling at configuration instead of a bifurcation edge case. I think this could be done by providing some additional receptacle metadata (as described below) and maybe modifying the target sampler with an AO precondition flag. This way it would be a per-sampler config option rather than a full generator/dataset config flag.

File: habitat/datasets/rearrange/rearrange_generator.py
Comment (Skylion007) [inline habitat/datasets/rearrange/rearrange_generator.py:437]:
```suggestion
            sampler = next(iter(self._obj_samplers.values()))
```
Guaranteed to be constant

File: habitat/datasets/rearrange/samplers.py
Comment (aclegg3) [inline habitat/datasets/rearrange/samplers.py:701]:
```suggestion
                            # because in this case all drawers must be closed.
                            # TODO: move this receptacle access logic to the ao_config files in a future refactor
```

**Review by aclegg3 — APPROVED** [review]
Thanks for refactoring to remove the hard-coded assumptions. 

These flag options are fine for now to unblock hab_suite_dev migration. 

@ASzot can merge support for multiple targets on top of this and we can revisit refactoring for more generality in the future.

**Pull request merged**
